### ISSUE TITLE
mapped intersection loses type information

### ISSUE THREAD

**Issue by kolodny** [Microsoft/TypeScript#30545]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:**
"mapped types", "overload",  "intersection"

**Code**

```ts
type ReturnObjects<T> = {
    [K in keyof T]: T[K] extends ((...args: any[]) => any) ?
    T[K] & Returnable<T[K]> :
    never
}

type Returnable<Fun extends (...args: any[]) => any> = {
    returnValue(val: ReturnType<Fun>): void;
}

interface I {
    f(): string;
    f(x: any): number;
}

let argsObj: ReturnObjects<I>;
argsObj.f.returnValue(1) // OK
argsObj.f.returnValue('1') // ERROR?!```
```

**Expected behavior:**

`argsObj.f.returnValue('1')` should not cause a type error

**Actual behavior:**

`argsObj.f.returnValue('1')` causes a type error

[**Playground Link**](https://www.typescriptlang.org/play/#src=type%20ReturnObjects%3CT%3E%20%3D%20%7B%0A%20%20%20%20%5BK%20in%20keyof%20T%5D%3A%20T%5BK%5D%20extends%20((...args%3A%20any%5B%5D)%20%3D%3E%20any)%20%3F%0A%20%20%20%20T%5BK%5D%20%26%20Returnable%3CT%5BK%5D%3E%20%3A%0A%20%20%20%20never%0A%7D%0A%0Atype%20Returnable%3CFun%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20any%3E%20%3D%20%7B%0A%20%20%20%20returnValue(val%3A%20ReturnType%3CFun%3E)%3A%20void%3B%0A%7D%0A%0Ainterface%20I%20%7B%0A%20%20%20%20f()%3A%20string%3B%0A%20%20%20%20f(x%3A%20any)%3A%20number%3B%0A%7D%0A%0Alet%20argsObj%3A%20ReturnObjects%3CI%3E%3B%0AargsObj.f.returnValue(1)%20%2F%2F%20OK%0AargsObj.f.returnValue('1')%20%2F%2F%20ERROR%3F!%0A%0A%0A%0Atype%20MapStrings%3CT%3E%20%3D%20%7B%0A%20%20%20%20%5BK%20in%20keyof%20T%5D%3A%20T%5BK%5D%20extends%20((...args%3A%20any%5B%5D)%20%3D%3E%20string)%20%3F%0A%20%20%20%20T%5BK%5D%20%26%20Returnable%3CT%5BK%5D%3E%20%3A%0A%20%20%20%20never%0A%7D%0A%0Atype%20MapNumbers%3CT%3E%20%3D%20%7B%0A%20%20%20%20%5BK%20in%20keyof%20T%5D%3A%20T%5BK%5D%20extends%20((...args%3A%20any%5B%5D)%20%3D%3E%20number)%20%3F%0A%20%20%20%20T%5BK%5D%20%26%20Returnable%3CT%5BK%5D%3E%20%3A%0A%20%20%20%20never%0A%7D%0A%0Alet%20tester%3A%20MapStrings%3CI%3E%20%26%20MapNumbers%3CI%3E%3B%0Atester.f()%20as%20string%3B%20%2F%2F%20OK%0Atester.f(1)%20as%20number%3B%20%2F%2F%20OK%0Atester.f.returnValue(123)%3B%20%2F%2F%20OK%0Atester.f.returnValue('123')%3B%20%2F%2F%20ERROR%3F!%0A%0A%2F%2F%20This%20seems%20to%20work%20when%20manually%20wiring%20up%20the%20union%20types%0A%0Alet%20a1%20%3D%20%7B%20returnValue%3A%20(arg%3A%20string)%20%3D%3E%20%7B%20%7D%20%7D%0Alet%20a2%20%3D%20%7B%20returnValue%3A%20(arg%3A%20number)%20%3D%3E%20%7B%20%7D%20%7D%0Alet%20manualUnion%3A%20typeof%20a1%20%26%20typeof%20a2%3B%0AmanualUnion.returnValue(3)%20%2F%2F%20OK%0AmanualUnion.returnValue('3')%20%2F%2F%20OK%0A%0A%2F%2F%20The%20function%20doesn't%20seem%20to%20be%20the%20cause%20since%20this%20works%20as%20well%0Ainterface%20A%20%7B%0A%20%20%20%20()%3A%20number%0A%20%20%20%20returnValue%3A%20(arg%3A%20number)%20%3D%3E%20void%0A%7D%0A%0Ainterface%20B%20%7B%0A%20%20%20%20(x%3A%20any)%3A%20string%0A%20%20%20%20returnValue%3A%20(arg%3A%20string)%20%3D%3E%20void%0A%7D%0A%0Alet%20ab%3A%20A%20%26%20B%3B%0Aab()%20as%20number%3B%20%2F%2F%20OK%0Aab(%7B%7D)%20as%20string%3B%20%2F%2F%20OK%0Aab.returnValue(1)%20%2F%2F%20OK%0Aab.returnValue('1')%20%2F%2F%20OK)

**Related Issues:** Couldn't find anything

Note that if I switch the order of the interface, it will only expect the other overload. I believe that typescript is overriding the previous value and not intersecting them

**Comment by jack-williams** [comment]
The problem here is that interface `I` contains an overloaded function that is passed to `ReturnType`. When inferring the return type only the last overload is chosen, as specified here #21496:

> When inferring from a type with multiple call signatures (such as the type of an overloaded function), inferences are made from the last signature (which, presumably, is the most permissive catch-all case). It is not possible to perform overload resolution based on a list of argument types (this would require us to support typeof for arbitrary expressions, as suggested in #6606, or something similar).

This is why the number input is ok, but the string input is not.
